<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of makeEcModel</title>
  <meta name="keywords" content="makeEcModel">
  <meta name="description" content="makeEcModel">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../../../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../../../index.html">Home</a> &gt;  <a href="#">src</a> &gt; <a href="#">geckomat</a> &gt; <a href="index.html">change_model</a> &gt; makeEcModel.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../../../index.html"><img alt="<" border="0" src="../../../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for src/geckomat/change_model&nbsp;<img alt=">" border="0" src="../../../right.png"></a></td></tr></table>-->

<h1>makeEcModel
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../../../up.png"></a></h2>
<div class="box"><strong>makeEcModel</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../../../up.png"></a></h2>
<div class="box"><strong>function [model, noUniprot] = makeEcModel(model, geckoLight, modelAdapter) </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../../../up.png"></a></h2>
<div class="fragment"><pre class="comment"> makeEcModel
   Expands a conventional genome-scale model (in RAVEN format) with enzyme
   information and prepares the reactions for integration of enzyme usage
   coefficients. This function contains all the steps that need to be done
   to get a basic ecModel, without incorporating any kcat values or
   constraints yet. This function should only have to be run once for a
   model.

 Input:
   model        a model in RAVEN format
   geckoLight   true if a simplified GECKO light model should be generated.
                (Optional, default is false).
   modelAdapter a loaded model adapter (Optional, will otherwise use the
                default model adapter).

 Ouput:
   model        an ecModel in GECKO 3 format, with a model.ec structure where
                enzyme and kcat information are stored. Protein pseudo-
                metabolites and their draw reactions are added to the model,
                but their usage is not yet implemented (due to absent kcat
                values at this stage).
   noUniprot    genes for which no information could be found in the
                Uniprot database

 The function goes through the following steps:
   1.  Remove gene associations from pseudoreactions.
   2.  Invert irreversible backwards reactions.
   3.  Correct 'rev' vector to match lb and ub vectors.
   4.  Convert to irreversible model (splits reversible reactions).
   5.  [Skipped with geckoLight:] Expand model to split reactions with
       'OR' in grRules (each reaction is then catalyzed by one enzyme
       (complex).
   6.  Make empty model.ec structure, that will contain enzyme and kcat
       information. One entry per reaction, where isoenzymes have multiple
       entries. This model.ec structure will later be populated with kcat
       values. For geckoLight the structure is different, where each
       reaction can have multiple isozymes.
   7.  Add enzyme information fields to model.ec structure: MW, sequence.
   8.  Populate model.ec structure with information from each reaction.
   9.  [Skipped with geckoLight:] Add proteins as pseudometabolites.
   10. Add prot_pool pseudometabolite.
   11. [Skipped with geckoLight:] Add usage reactions for the protein
       pseudometabolites, replenishing from the protein pool (default, can
       be changed to consider proteomics data at later stage)
   12. Add protein pool reaction, without upper bound.

   Note that while protein pseudometabolites, draw &amp; pool reactions might
   be added to the model, the enzyme usage is not yet incorporated in each
   metabolic reaction, so enzymes will not be used. applyKcatConstraints
   incorporates protein pseudometabolites in reactions as enzyme usages by
   applying the specified kcats as constraints.

The EC structure looks as follows
 Attributes:
   geckoLight: 0 if full model, 1 if light
         rxns: reaction identifiers that correspond to model.rxns
         kcat: kcat values - not set here
       source: specifies where the kcats come from - not set here
        notes: notes that can be set by downstream functions - not set
               here
      eccodes: enzyme codes for each enzyme - not set here
        genes: the genes involved in the kcats - not necessarily the
               same as model.genes, since some genes may not be found in
               databases etc.
      enzymes: Uniprot protein identifiers for the genes
           mw: molecular weights of the enzymes
     sequence: sequence of the genes/enzymes
        concs: concentrations of the enzymes - not set here
    rxnEnzMat: matrix of enzymes and rxns

 The full model is split on all ORs in the GPRs, meaning that the
 reactions will be duplicated for each isozyme. Only the rxns with genes
 are added. The fields rxns, eccodes, kcat, source and notes will
 therefore have one entry per reaction. The fields genes, enzymes, mw,
 sequence and concs will have one entry per gene. The rxnEnzMat is a
 matrix with reactions and genes, mapping which genes are connected to
 which reaction (where isozymes have different reactions).

 The light model works a bit differently. The model has the same number of
 rxns as the original model, but expanded since it is reversible + one the
 extra prot maintenance rxn and one extra prot_pool rxn. However, the ec
 fields rxns, eccodes, kcat, source and notes are duplicated for each
 isozyme, sorted the same way as model.rxns. So, in model.ec.rxns, the
 same reaction will appear several times after one another, one entry per
 izozyme, with corresponding values for that isozyme. These fields
 therefore have the same length as for the full model. The fields genes,
 enzymes, mw, sequence and concs are the same here as in the full model.
 The rxnEnzMat maps the model.ec.rxns entries to genes and is therefore of
 the same size as for the full model.

 Usage:
   [model, noUniprot] = makeEcModel(model, geckoLight, modelAdapter)</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../../../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../../../matlabicon.gif)">
</ul>
This function is called by:
<ul style="list-style-image:url(../../../matlabicon.gif)">
</ul>
<!-- crossreference -->

<h2><a name="_subfunctions"></a>SUBFUNCTIONS <a href="#_top"><img alt="^" border="0" src="../../../up.png"></a></h2>
<ul style="list-style-image:url(../../../matlabicon.gif)">
<li><a href="#_sub1" class="code">function indexes2check = findPotentialErrors(grRules,embedded,model)</a></li></ul>

<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../../../up.png"></a></h2>
<div class="fragment"><pre>0001 <a name="_sub0" href="#_subfunctions" class="code">function [model, noUniprot] = makeEcModel(model, geckoLight, modelAdapter)</a>
0002 <span class="comment">% makeEcModel</span>
0003 <span class="comment">%   Expands a conventional genome-scale model (in RAVEN format) with enzyme</span>
0004 <span class="comment">%   information and prepares the reactions for integration of enzyme usage</span>
0005 <span class="comment">%   coefficients. This function contains all the steps that need to be done</span>
0006 <span class="comment">%   to get a basic ecModel, without incorporating any kcat values or</span>
0007 <span class="comment">%   constraints yet. This function should only have to be run once for a</span>
0008 <span class="comment">%   model.</span>
0009 <span class="comment">%</span>
0010 <span class="comment">% Input:</span>
0011 <span class="comment">%   model        a model in RAVEN format</span>
0012 <span class="comment">%   geckoLight   true if a simplified GECKO light model should be generated.</span>
0013 <span class="comment">%                (Optional, default is false).</span>
0014 <span class="comment">%   modelAdapter a loaded model adapter (Optional, will otherwise use the</span>
0015 <span class="comment">%                default model adapter).</span>
0016 <span class="comment">%</span>
0017 <span class="comment">% Ouput:</span>
0018 <span class="comment">%   model        an ecModel in GECKO 3 format, with a model.ec structure where</span>
0019 <span class="comment">%                enzyme and kcat information are stored. Protein pseudo-</span>
0020 <span class="comment">%                metabolites and their draw reactions are added to the model,</span>
0021 <span class="comment">%                but their usage is not yet implemented (due to absent kcat</span>
0022 <span class="comment">%                values at this stage).</span>
0023 <span class="comment">%   noUniprot    genes for which no information could be found in the</span>
0024 <span class="comment">%                Uniprot database</span>
0025 <span class="comment">%</span>
0026 <span class="comment">% The function goes through the following steps:</span>
0027 <span class="comment">%   1.  Remove gene associations from pseudoreactions.</span>
0028 <span class="comment">%   2.  Invert irreversible backwards reactions.</span>
0029 <span class="comment">%   3.  Correct 'rev' vector to match lb and ub vectors.</span>
0030 <span class="comment">%   4.  Convert to irreversible model (splits reversible reactions).</span>
0031 <span class="comment">%   5.  [Skipped with geckoLight:] Expand model to split reactions with</span>
0032 <span class="comment">%       'OR' in grRules (each reaction is then catalyzed by one enzyme</span>
0033 <span class="comment">%       (complex).</span>
0034 <span class="comment">%   6.  Make empty model.ec structure, that will contain enzyme and kcat</span>
0035 <span class="comment">%       information. One entry per reaction, where isoenzymes have multiple</span>
0036 <span class="comment">%       entries. This model.ec structure will later be populated with kcat</span>
0037 <span class="comment">%       values. For geckoLight the structure is different, where each</span>
0038 <span class="comment">%       reaction can have multiple isozymes.</span>
0039 <span class="comment">%   7.  Add enzyme information fields to model.ec structure: MW, sequence.</span>
0040 <span class="comment">%   8.  Populate model.ec structure with information from each reaction.</span>
0041 <span class="comment">%   9.  [Skipped with geckoLight:] Add proteins as pseudometabolites.</span>
0042 <span class="comment">%   10. Add prot_pool pseudometabolite.</span>
0043 <span class="comment">%   11. [Skipped with geckoLight:] Add usage reactions for the protein</span>
0044 <span class="comment">%       pseudometabolites, replenishing from the protein pool (default, can</span>
0045 <span class="comment">%       be changed to consider proteomics data at later stage)</span>
0046 <span class="comment">%   12. Add protein pool reaction, without upper bound.</span>
0047 <span class="comment">%</span>
0048 <span class="comment">%   Note that while protein pseudometabolites, draw &amp; pool reactions might</span>
0049 <span class="comment">%   be added to the model, the enzyme usage is not yet incorporated in each</span>
0050 <span class="comment">%   metabolic reaction, so enzymes will not be used. applyKcatConstraints</span>
0051 <span class="comment">%   incorporates protein pseudometabolites in reactions as enzyme usages by</span>
0052 <span class="comment">%   applying the specified kcats as constraints.</span>
0053 <span class="comment">%</span>
0054 <span class="comment">%The EC structure looks as follows</span>
0055 <span class="comment">% Attributes:</span>
0056 <span class="comment">%   geckoLight: 0 if full model, 1 if light</span>
0057 <span class="comment">%         rxns: reaction identifiers that correspond to model.rxns</span>
0058 <span class="comment">%         kcat: kcat values - not set here</span>
0059 <span class="comment">%       source: specifies where the kcats come from - not set here</span>
0060 <span class="comment">%        notes: notes that can be set by downstream functions - not set</span>
0061 <span class="comment">%               here</span>
0062 <span class="comment">%      eccodes: enzyme codes for each enzyme - not set here</span>
0063 <span class="comment">%        genes: the genes involved in the kcats - not necessarily the</span>
0064 <span class="comment">%               same as model.genes, since some genes may not be found in</span>
0065 <span class="comment">%               databases etc.</span>
0066 <span class="comment">%      enzymes: Uniprot protein identifiers for the genes</span>
0067 <span class="comment">%           mw: molecular weights of the enzymes</span>
0068 <span class="comment">%     sequence: sequence of the genes/enzymes</span>
0069 <span class="comment">%        concs: concentrations of the enzymes - not set here</span>
0070 <span class="comment">%    rxnEnzMat: matrix of enzymes and rxns</span>
0071 <span class="comment">%</span>
0072 <span class="comment">% The full model is split on all ORs in the GPRs, meaning that the</span>
0073 <span class="comment">% reactions will be duplicated for each isozyme. Only the rxns with genes</span>
0074 <span class="comment">% are added. The fields rxns, eccodes, kcat, source and notes will</span>
0075 <span class="comment">% therefore have one entry per reaction. The fields genes, enzymes, mw,</span>
0076 <span class="comment">% sequence and concs will have one entry per gene. The rxnEnzMat is a</span>
0077 <span class="comment">% matrix with reactions and genes, mapping which genes are connected to</span>
0078 <span class="comment">% which reaction (where isozymes have different reactions).</span>
0079 <span class="comment">%</span>
0080 <span class="comment">% The light model works a bit differently. The model has the same number of</span>
0081 <span class="comment">% rxns as the original model, but expanded since it is reversible + one the</span>
0082 <span class="comment">% extra prot maintenance rxn and one extra prot_pool rxn. However, the ec</span>
0083 <span class="comment">% fields rxns, eccodes, kcat, source and notes are duplicated for each</span>
0084 <span class="comment">% isozyme, sorted the same way as model.rxns. So, in model.ec.rxns, the</span>
0085 <span class="comment">% same reaction will appear several times after one another, one entry per</span>
0086 <span class="comment">% izozyme, with corresponding values for that isozyme. These fields</span>
0087 <span class="comment">% therefore have the same length as for the full model. The fields genes,</span>
0088 <span class="comment">% enzymes, mw, sequence and concs are the same here as in the full model.</span>
0089 <span class="comment">% The rxnEnzMat maps the model.ec.rxns entries to genes and is therefore of</span>
0090 <span class="comment">% the same size as for the full model.</span>
0091 <span class="comment">%</span>
0092 <span class="comment">% Usage:</span>
0093 <span class="comment">%   [model, noUniprot] = makeEcModel(model, geckoLight, modelAdapter)</span>
0094 
0095 <span class="keyword">if</span> nargin&lt;2
0096     geckoLight=false;
0097 <span class="keyword">elseif</span> ~islogical(geckoLight) &amp;&amp; ~(geckoLight == 0) &amp;&amp; ~(geckoLight == 1)
0098     error(<span class="string">'geckoLight should be either true or false'</span>)
0099 <span class="keyword">end</span>
0100 
0101 <span class="keyword">if</span> nargin &lt; 3 || isempty(modelAdapter)
0102     modelAdapter = ModelAdapterManager.getDefaultAdapter();
0103     <span class="keyword">if</span> isempty(modelAdapter)
0104         error(<span class="string">'Either send in a modelAdapter or set the default model adapter in the ModelAdapterManager.'</span>)
0105     <span class="keyword">end</span>
0106 <span class="keyword">end</span>
0107 params = modelAdapter.getParameters();
0108 compartmentID = strcmp(model.compNames,params.enzyme_comp);
0109 <span class="keyword">if</span> ~any(compartmentID)
0110     error([<span class="string">'Compartment '</span> params.enzyme_comp <span class="string">' (specified in params.enzyme_comp) '</span><span class="keyword">...</span>
0111            <span class="string">'cannot be found in model.compNames'</span>])
0112 <span class="keyword">end</span>
0113 compartmentID = model.comps(compartmentID);
0114 
0115 <span class="keyword">if</span> geckoLight
0116     ec.geckoLight=true;
0117 <span class="keyword">else</span>
0118     ec.geckoLight=false;
0119 <span class="keyword">end</span>
0120 
0121 <span class="comment">%Check if model is in RAVEN format</span>
0122 <span class="keyword">if</span> any(isfield(model,{<span class="string">'rules'</span>,<span class="string">'modelID'</span>}))
0123     error([<span class="string">'The model is likely loaded using COBRA Toolbox readCbModel(). Instead, use '</span> <span class="keyword">...</span>
0124            <span class="string">'RAVEN Toolbox importModel(). Alternatively, you can also convert the '</span>, <span class="keyword">...</span>
0125            <span class="string">'model in MATLAB using ravenCobraWrapper().'</span>])
0126 <span class="keyword">end</span>
0127 
0128 <span class="comment">%Check for conflicting reaction and metabolite identifiers</span>
0129 conflictId = startsWith(model.mets,<span class="string">'prot_'</span>);
0130 <span class="keyword">if</span> any(conflictId)
0131     error(<span class="string">'The identifiers in model.mets are not allowed to start with ''prot_''.'</span>)
0132 <span class="keyword">end</span>
0133 conflictId = startsWith(model.rxns,{<span class="string">'usage_prot_'</span>,<span class="string">'prot_pool'</span>}) | <span class="keyword">...</span>
0134              endsWith(model.rxns,<span class="string">'_REV'</span>) | <span class="keyword">...</span>
0135              ~cellfun(@isempty,regexp(model.rxns,<span class="string">'_EXP_\d+$'</span>));
0136 <span class="keyword">if</span> any(conflictId)
0137     error([<span class="string">'The identifiers in model.rxns are not allowed to start with ''usage_prot'' '</span> <span class="keyword">...</span>
0138            <span class="string">'or ''prot_pool'', or end with ''_REV'' or ''_EXP_[digit]''.'</span>])
0139 <span class="keyword">end</span>
0140 
0141 uniprotDB = loadDatabases(<span class="string">'both'</span>, modelAdapter);
0142 uniprotDB = uniprotDB.uniprot;
0143 
0144 <span class="comment">%1: Remove gene rules from pseudoreactions (if any):</span>
0145 <span class="keyword">for</span> i = 1:length(model.rxns)
0146     <span class="keyword">if</span> endsWith(model.rxnNames{i},<span class="string">' pseudoreaction'</span>)
0147         model.grRules{i}      = <span class="string">''</span>;
0148         model.rxnGeneMat(i,:) = zeros(1,length(model.genes));
0149     <span class="keyword">end</span>
0150 <span class="keyword">end</span>
0151 
0152 <span class="comment">%2: Swap direction of reactions that are defined to only carry negative flux</span>
0153 to_swap=model.lb &lt; 0 &amp; model.ub == 0;
0154 model.S(:,to_swap)=-model.S(:,to_swap);
0155 model.ub(to_swap)=-model.lb(to_swap);
0156 model.lb(to_swap)=0;
0157 
0158 <span class="comment">%3: Correct rev vector: true if LB &lt; 0 &amp; UB &gt; 0, or it is an exchange reaction:</span>
0159 model.rev = false(size(model.rxns));
0160 <span class="keyword">for</span> i = 1:length(model.rxns)
0161     <span class="keyword">if</span> (model.lb(i) &lt; 0 &amp;&amp; model.ub(i) &gt; 0) || sum(model.S(:,i) ~= 0) == 1
0162         model.rev(i) = true;
0163     <span class="keyword">end</span>
0164 <span class="keyword">end</span>
0165 
0166 <span class="comment">%4: Make irreversible model (appends _REV to reaction IDs to indicate reverse</span>
0167 <span class="comment">%reactions)</span>
0168 [~,exchRxns] = getExchangeRxns(model);
0169 nonExchRxns = model.rxns;
0170 nonExchRxns(exchRxns) = [];
0171 model=convertToIrrev(model, nonExchRxns);
0172 
0173 <span class="comment">%5: Expand model, to separate isoenzymes (appends _EXP_* to reaction IDs to</span>
0174 <span class="comment">%indicate duplication)</span>
0175 <span class="keyword">if</span> ~geckoLight
0176     model=expandModel(model);
0177 <span class="keyword">end</span>
0178 <span class="comment">% Sort reactions, so that reversible and isoenzymic reactions are kept near</span>
0179 <span class="keyword">if</span> ~geckoLight
0180     model=sortIdentifiers(model);
0181 <span class="keyword">end</span>
0182 
0183 <span class="comment">%6: Make ec-extension structure, one for gene-associated reaction.</span>
0184 <span class="comment">%   The structure is different for light and full models</span>
0185 rxnWithGene  = find(sum(model.rxnGeneMat,2));
0186 <span class="keyword">if</span> ~geckoLight
0187     ec.rxns      = model.rxns(rxnWithGene);
0188     emptyCell    = cell(numel(rxnWithGene),1);
0189     emptyCell(:) = {<span class="string">''</span>};
0190     emptyVect    = zeros(numel(rxnWithGene),1);
0191     
0192     ec.kcat      = emptyVect;
0193     ec.source    = emptyCell; <span class="comment">% Strings, like 'dlkcat', 'manual', 'brenda', etc.</span>
0194     ec.notes     = emptyCell; <span class="comment">% Additional comments</span>
0195     ec.eccodes   = emptyCell;
0196     ec.concs     = emptyVect;
0197 <span class="keyword">else</span>
0198     <span class="comment">%Different strategy for GECKO light: Each reaction can exist multiple times in</span>
0199     <span class="comment">%ec.rxns and similar fields - one time per isozyme. The number of copies is</span>
0200     <span class="comment">%the number of ORs in the GPR + 1</span>
0201     numOrs = count(model.grRules(rxnWithGene), <span class="string">' or '</span>);
0202     cpys = numOrs + 1;
0203     prevNumRxns = length(numOrs);
0204     cpyIndices = repelem(rxnWithGene, cpys).'; <span class="comment">%.' only matter when number of rxns is 1</span>
0205     <span class="comment">%loop through and add a prefix with an isozyme index to the rxns</span>
0206     <span class="comment">%we just give a fixed-length number as prefix, and assume that 999 is enough</span>
0207     tmpRxns = model.rxns(cpyIndices); <span class="comment">%now they have no prefix</span>
0208     newRxns = tmpRxns;
0209     
0210     <span class="comment">%add the prefix</span>
0211     nextIndex = 1;
0212     <span class="keyword">for</span> i = 1:numel(model.rxns)
0213         localRxnIndex = 1;
0214         <span class="keyword">if</span> nextIndex &lt;= length(tmpRxns) &amp;&amp; strcmp(model.rxns(i), tmpRxns(nextIndex))
0215             <span class="keyword">while</span> true
0216                 tmp = compose(<span class="string">'%03d_'</span>,localRxnIndex);
0217                 newRxns{nextIndex} = [tmp{1} tmpRxns{nextIndex}];
0218                 localRxnIndex = localRxnIndex + 1;
0219                 <span class="keyword">if</span> (localRxnIndex &gt;= 1000)
0220                     error(<span class="string">'Increase index size to 10000 - error in the code.'</span>); <span class="comment">%this should never happen, we don't have &gt; 999 isozymes</span>
0221                 <span class="keyword">end</span>
0222                 nextIndex = nextIndex + 1;
0223                 <span class="keyword">if</span>  nextIndex &gt; length(tmpRxns) || ~strcmp(model.rxns(i), tmpRxns(nextIndex))
0224                     <span class="keyword">break</span>;
0225                 <span class="keyword">end</span>
0226             <span class="keyword">end</span>
0227         <span class="keyword">end</span>
0228     <span class="keyword">end</span>
0229 
0230     ec.rxns      = newRxns;
0231     
0232     emptyCell    = cell(numel(ec.rxns),1);
0233     emptyCell(:) = {<span class="string">''</span>};
0234     emptyVect    = zeros(numel(ec.rxns),1);
0235 
0236     ec.kcat      = emptyVect;
0237     ec.source    = emptyCell; <span class="comment">% Strings, like 'dlkcat', 'manual', 'brenda', etc.</span>
0238     ec.notes     = emptyCell; <span class="comment">% Additional comments</span>
0239     ec.eccodes   = emptyCell;
0240     ec.concs     = emptyVect;
0241 <span class="keyword">end</span>
0242     
0243 <span class="comment">%7: Gather enzyme information via UniprotDB</span>
0244 uniprotCompatibleGenes = modelAdapter.getUniprotCompatibleGenes(model.genes);
0245 [Lia,Locb] = ismember(uniprotCompatibleGenes,uniprotDB.genes);
0246 
0247 uniprot = modelAdapter.getUniprotIDsFromTable(uniprotCompatibleGenes);
0248 <span class="keyword">if</span> ~isequal(uniprot,uniprotCompatibleGenes)
0249     uniprot(cellfun(@isempty,uniprot)) = {<span class="string">''</span>};
0250     [Lia,Locb] = ismember(uniprot,uniprotDB.ID);
0251 <span class="keyword">end</span>
0252 noUniprot  = uniprotCompatibleGenes(~Lia);
0253 <span class="keyword">if</span> ~isempty(noUniprot)
0254     warning([<span class="string">'The '</span> num2str(numel(noUniprot)) <span class="string">' gene(s) reported in noUniprot cannot '</span><span class="keyword">...</span>
0255           <span class="string">'be found in data/uniprot.tsv, these will not be enzyme-constrained. '</span><span class="keyword">...</span>
0256           <span class="string">'If you intend to use different Uniprot data (e.g. from a different proteome '</span><span class="keyword">...</span>
0257           <span class="string">'identifier), make sure you first delete the existing data/uniprot.tsv file.'</span>])
0258 <span class="keyword">end</span>
0259 ec.genes        = model.genes(Lia); <span class="comment">%Will often be duplicate of model.genes, but is done here to prevent issues when it is not.</span>
0260 ec.enzymes      = uniprotDB.ID(Locb(Lia));
0261 ec.mw           = uniprotDB.MW(Locb(Lia));
0262 ec.sequence     = uniprotDB.seq(Locb(Lia));
0263 <span class="comment">%Additional info</span>
0264 ec.concs        = nan(numel(ec.genes),1); <span class="comment">% To be filled with proteomics data when available</span>
0265 
0266 <span class="comment">%8: Only parse rxns associated to genes</span>
0267 <span class="keyword">if</span> ~geckoLight
0268     ec.rxnEnzMat = zeros(numel(rxnWithGene),numel(ec.genes)); <span class="comment">% Non-zeros will indicate the number of subunits</span>
0269     <span class="keyword">for</span> r=1:numel(rxnWithGene)
0270         rxnGenes   = model.genes(find(model.rxnGeneMat(rxnWithGene(r),:)));
0271         [~,locEnz] = ismember(rxnGenes,ec.genes); <span class="comment">% Could also parse directly from rxnGeneMat, but some genes might be missing from Uniprot DB</span>
0272         <span class="keyword">if</span> locEnz ~= 0
0273             ec.rxnEnzMat(r,locEnz) = 1; <span class="comment">%Assume 1 copy per subunit or enzyme, can be modified later</span>
0274         <span class="keyword">end</span>
0275     <span class="keyword">end</span>
0276 <span class="keyword">else</span>
0277     <span class="comment">%For light models, we need to split up all GPRs</span>
0278     ec.rxnEnzMat = zeros(numel(ec.rxns),numel(ec.genes)); <span class="comment">% Non-zeros will indicate the number of subunits</span>
0279     nextIndex = 1;
0280     <span class="comment">%For full model generation, the GPRs are controlled in expandModel, but</span>
0281     <span class="comment">%here we need to make an explicit format check</span>
0282     indexes2check = <a href="#_sub1" class="code" title="subfunction indexes2check = findPotentialErrors(grRules,embedded,model)">findPotentialErrors</a>(model.grRules,false,model);
0283     <span class="keyword">if</span> ~isempty(indexes2check) 
0284         disp(<span class="string">'For Human-GEM, these reactions can be corrected using simplifyGrRules.'</span>);
0285     <span class="keyword">end</span>
0286     
0287     <span class="keyword">for</span> i=1:prevNumRxns
0288         <span class="comment">%ind is the index in the model, not to confuse with the index in the ec struct (i),</span>
0289         <span class="comment">%which only contains reactions with GPRs.</span>
0290         ind = rxnWithGene(i); 
0291         <span class="comment">%Get rid of all '(' and ')' since I'm not looking at complex stuff</span>
0292         <span class="comment">%anyways</span>
0293         geneString=model.grRules{ind};
0294         geneString=strrep(geneString,<span class="string">'('</span>,<span class="string">''</span>);
0295         geneString=strrep(geneString,<span class="string">')'</span>,<span class="string">''</span>);
0296         geneString=strrep(geneString,<span class="string">' or '</span>,<span class="string">';'</span>);
0297         
0298         <span class="keyword">if</span> (numOrs(i) == 0)
0299             geneNames = {geneString};
0300         <span class="keyword">else</span>
0301             <span class="comment">%Split the string into gene names</span>
0302             geneNames=regexp(geneString,<span class="string">';'</span>,<span class="string">'split'</span>);
0303         <span class="keyword">end</span>
0304         
0305         <span class="comment">%Now loop through the isozymes and set the rxnGeneMat</span>
0306         <span class="keyword">for</span> j = 1:length(geneNames)
0307             <span class="comment">%Find the gene in the gene list If ' and ' relationship, first</span>
0308             <span class="comment">%split the genes</span>
0309             fnd = strfind(geneNames{j},<span class="string">' and '</span>);
0310             <span class="keyword">if</span> ~isempty(fnd)
0311                 andGenes=regexp(geneNames{j},<span class="string">' and '</span>,<span class="string">'split'</span>);
0312                 ec.rxnEnzMat(nextIndex,ismember(ec.genes,andGenes)) = 1; <span class="comment">%should be subunit stoichoimetry</span>
0313             <span class="keyword">else</span>
0314                 ec.rxnEnzMat(nextIndex,ismember(ec.genes,geneNames(j)))=1;<span class="comment">%should be subunit stoichoimetry</span>
0315             <span class="keyword">end</span>
0316             nextIndex = nextIndex + 1;
0317         <span class="keyword">end</span>
0318     <span class="keyword">end</span>
0319 <span class="keyword">end</span>
0320 
0321 <span class="comment">%9: Add proteins as pseudometabolites</span>
0322 <span class="keyword">if</span> ~geckoLight
0323     [proteinMets.mets, uniprotSortId] = sort(ec.enzymes);
0324     proteinMets.mets         = strcat(<span class="string">'prot_'</span>,proteinMets.mets);
0325     proteinMets.metNames     = proteinMets.mets;
0326     proteinMets.compartments = compartmentID;
0327     <span class="keyword">if</span> isfield(model,<span class="string">'metMiriams'</span>)
0328         proteinMets.metMiriams   = repmat({struct(<span class="string">'name'</span>,{{<span class="string">'sbo'</span>}},<span class="string">'value'</span>,{{<span class="string">'SBO:0000252'</span>}})},numel(proteinMets.mets),1);
0329     <span class="keyword">end</span>
0330     proteinMets.metNotes     = repmat({<span class="string">'Enzyme-usage pseudometabolite'</span>},numel(proteinMets.mets),1);
0331     model = addMets(model,proteinMets);
0332 <span class="keyword">end</span>
0333 
0334 <span class="comment">%10: Add protein pool pseudometabolite</span>
0335 pool.mets         = <span class="string">'prot_pool'</span>;
0336 pool.metNames     = pool.mets;
0337 pool.compartments = compartmentID;
0338 pool.metNotes     = <span class="string">'Enzyme-usage protein pool'</span>;
0339 model = addMets(model,pool);
0340 
0341 <span class="comment">%11: Add protein usage reactions.</span>
0342 <span class="keyword">if</span> ~geckoLight
0343     usageRxns.rxns            = strcat(<span class="string">'usage_'</span>,proteinMets.mets);
0344     usageRxns.rxnNames        = usageRxns.rxns;
0345     usageRxns.mets            = cell(numel(usageRxns.rxns),1);
0346     usageRxns.stoichCoeffs    = cell(numel(usageRxns.rxns),1);
0347     <span class="keyword">for</span> i=1:numel(usageRxns.mets)
0348         usageRxns.mets{i}         = {proteinMets.mets{i}, <span class="string">'prot_pool'</span>};
0349         usageRxns.stoichCoeffs{i} = [-1,1];
0350     <span class="keyword">end</span>
0351     usageRxns.lb              = zeros(numel(usageRxns.rxns),1) - 1000;
0352     usageRxns.ub              = zeros(numel(usageRxns.rxns),1);
0353     usageRxns.grRules         = ec.genes(uniprotSortId);
0354     model = addRxns(model,usageRxns);
0355 <span class="keyword">end</span>
0356 
0357 <span class="comment">%12: Add protein pool reaction (with open UB)</span>
0358 poolRxn.rxns            = <span class="string">'prot_pool_exchange'</span>;
0359 poolRxn.rxnNames        = poolRxn.rxns;
0360 poolRxn.mets            = {<span class="string">'prot_pool'</span>};
0361 poolRxn.stoichCoeffs    = {-1};
0362 poolRxn.lb              = -1000;
0363 poolRxn.ub              = 0;
0364 model = addRxns(model,poolRxn);
0365 
0366 model.ec=ec;
0367 <span class="keyword">end</span>
0368 
0369 <span class="comment">%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%</span>
0370 <span class="comment">%Function that gets the model field grRules and returns the indexes of the</span>
0371 <span class="comment">%rules in which the pattern &quot;) and (&quot; is present.</span>
0372 <span class="comment">%Copied from standardizeGrRules</span>
0373 <span class="comment">% TODO: Make this an accessible function in a separate file in RAVEN and remove this</span>
0374 <span class="comment">%implementation.</span>
0375 <a name="_sub1" href="#_subfunctions" class="code">function indexes2check = findPotentialErrors(grRules,embedded,model)</a>
0376 indxs_l       = find(~cellfun(@isempty,strfind(grRules,<span class="string">') and ('</span>)));
0377 indxs_l_L     = find(~cellfun(@isempty,strfind(grRules,<span class="string">') and'</span>)));
0378 indxs_l_R     = find(~cellfun(@isempty,strfind(grRules,<span class="string">'and ('</span>)));
0379 indexes2check = vertcat(indxs_l,indxs_l_L,indxs_l_R);
0380 indexes2check = unique(indexes2check);
0381 
0382 <span class="keyword">if</span> ~isempty(indexes2check)
0383     
0384     <span class="keyword">if</span> embedded
0385         EM = <span class="string">'Potentially problematic &quot;) AND (&quot; in the grRules for reaction(s): '</span>;
0386         dispEM(EM,false,model.rxns(indexes2check),true)
0387     <span class="keyword">else</span>
0388         STR = <span class="string">'Potentially problematic &quot;) AND (&quot;, &quot;) AND&quot; or &quot;AND (&quot;relat'</span>;
0389         STR = [STR,<span class="string">'ionships found in\n\n'</span>];
0390         <span class="keyword">for</span> i=1:length(indexes2check)
0391             index = indexes2check(i);
0392             STR = [STR <span class="string">'  - grRule #'</span> model.rxns{index} <span class="string">': '</span> grRules{index} <span class="string">'\n'</span>];
0393         <span class="keyword">end</span>
0394         STR = [STR,<span class="string">'\n This kind of relationships should only be present '</span>];
0395         STR = [STR,<span class="string">'in  reactions catalysed by complexes of isoenzymes e'</span>];
0396         STR = [STR,<span class="string">'.g.\n\n  - (G1 or G2) and (G3 or G4)\n\n For these c'</span>];
0397         STR = [STR,<span class="string">'ases modify the grRules manually, writing all the po'</span>];
0398         STR = [STR,<span class="string">'ssible combinations e.g.\n\n  - (G1 and G3) or (G1 a'</span>];
0399         STR = [STR,<span class="string">'nd G4) or (G2 and G3) or (G2 and G4)\n\n For other c'</span>];
0400         STR = [STR,<span class="string">'ases modify the correspondent grRules avoiding:\n\n '</span>];
0401         STR = [STR,<span class="string">' 1) Overall container brackets, e.g.\n        &quot;(G1 a'</span>];
0402         STR = [STR,<span class="string">'nd G2)&quot; should be &quot;G1 and G2&quot;\n\n  2) Single unit en'</span>];
0403         STR = [STR,<span class="string">'zymes enclosed into brackets, e.g.\n        &quot;(G1)&quot; s'</span>];
0404         STR = [STR,<span class="string">'hould be &quot;G1&quot;\n\n  3) The use of uppercases for logi'</span>];
0405         STR = [STR,<span class="string">'cal operators, e.g.\n        &quot;G1 OR G2&quot; should be &quot;G'</span>];
0406         STR = [STR,<span class="string">'1 or G2&quot;\n\n  4) Unbalanced brackets, e.g.\n        '</span>];
0407         STR = [STR,<span class="string">'&quot;((G1 and G2) or G3&quot; should be &quot;(G1 and G2) or G3&quot;\n'</span>];
0408         warning(sprintf(STR))
0409     <span class="keyword">end</span>
0410 <span class="keyword">end</span>
0411 <span class="keyword">end</span>
0412</pre></div>
<hr><address>Generated by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>